### PR TITLE
Add middleware to address transitional location of tenantId field

### PR DESCRIPTION
**Pull request by benbrown**
This is the C# version of [this Javascript PR](https://github.com/Microsoft/botbuilder-js/pull/886)

## Description
This change handles the in-transition location of the tenant ID field for MS Teams.

## Specific Changes
Load a middleware in the constructor for BotFrameworkAdapter that will look for channelData.tenant.id and copy the value to conversation.tenantId for use in createConversation.

## Testing
* Connect a bot to MS Teams, then use `adapter.createConversation` to create a DM with a user.

### CODE REVIEW

**Comment by cleemullins** [comment]
```
No issues found in Microsoft.Bot.Builder.dll
No issues found in Microsoft.Bot.Builder.AI.Luis.dll
No issues found in Microsoft.Bot.Builder.AI.QnA.dll
No issues found in Microsoft.Bot.Builder.ApplicationInsights.dll
No issues found in Microsoft.Bot.Builder.Azure.dll
No issues found in Microsoft.Bot.Builder.Dialogs.dll
No issues found in Microsoft.Bot.Builder.TemplateManager.dll
No issues found in Microsoft.Bot.Configuration.dll
No issues found in Microsoft.Bot.Connector.dll
No issues found in Microsoft.Bot.Schema.dll

```

**Comment by gabog** [comment]
@benbrown, I updated the middleware to also check for ChannelId = MSTeams and also added some unit tests for it.
@cleemullins, I noticed some issues with the adapter code that I fixed too (please take a look if you have time). These changes include: 
- naming conventons
- typos in documentation and variables 
- changed some validations to throw NullReferenceException instead of ArgumentNullExceptions (IMHO, ArgumentNullException was not the right exception to throw and we were using the argName parameter to pass an error message)

**Comment by cleemullins** [comment]
```
No issues found in Microsoft.Bot.Builder.dll
No issues found in Microsoft.Bot.Builder.AI.Luis.dll
No issues found in Microsoft.Bot.Builder.AI.QnA.dll
No issues found in Microsoft.Bot.Builder.ApplicationInsights.dll
No issues found in Microsoft.Bot.Builder.Azure.dll
No issues found in Microsoft.Bot.Builder.Dialogs.dll
No issues found in Microsoft.Bot.Builder.TemplateManager.dll
No issues found in Microsoft.Bot.Configuration.dll
No issues found in Microsoft.Bot.Connector.dll
No issues found in Microsoft.Bot.Schema.dll

```

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 1]
```diff
@@ -393,12 +399,12 @@ public async Task DeleteConversationMemberAsync(ITurnContext turnContext, string
         {
             if (turnContext.Activity.Conversation == null)
             {
-                throw new ArgumentNullException("BotFrameworkAdapter.deleteConversationMember(): missing conversation");
+                throw new NullReferenceException("BotFrameworkAdapter.deleteConversationMember(): missing conversation");
             }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Please don't change this. That's changing behaviorial semantics and shouldn't be done here.

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 2]
```diff
@@ -393,12 +399,12 @@ public async Task DeleteConversationMemberAsync(ITurnContext turnContext, string
...
 
             if (string.IsNullOrWhiteSpace(turnContext.Activity.Conversation.Id))
             {
-                throw new ArgumentNullException("BotFrameworkAdapter.deleteConversationMember(): missing conversation.id");
+                throw new NullReferenceException("BotFrameworkAdapter.deleteConversationMember(): missing conversation.id");
             }
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: All of these shouldn't change.

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 3]
```diff
@@ -932,5 +939,25 @@ private async Task<MicrosoftAppCredentials> GetAppCredentialsAsync(string appId,
             _appCredentialMap[appId] = appCredentials;
             return appCredentials;
         }
+
+        /// <summary>
+        /// Middleware to assign tenantId from channelData to Conversation.TenantId.
+        /// </summary>
+        internal class TenantIdWorkaroundForTeamsMiddleware : IMiddleware
+        {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Please add a far more detailed comment here. Why is this needed? When can we deprecate it?

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 4]
```diff
@@ -932,5 +939,25 @@ private async Task<MicrosoftAppCredentials> GetAppCredentialsAsync(string appId,
...
+            public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = default(CancellationToken))
+            {
+                if (turnContext.Activity.ChannelId == Channels.Msteams && turnContext.Activity.Conversation != null && string.IsNullOrEmpty(turnContext.Activity.Conversation.TenantId))
+                {
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: I would update the "==" compare against Channels.Msteams to be culture invariant.

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 3 above)
Comment (gabog) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Added additional documentation to describe why and when can this be removed.

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 1 above)
Comment (gabog) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Reverted

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 2 above)
Comment (gabog) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: reverted

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 4 above)
Comment (gabog) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Addressed

File: tests/Microsoft.Bot.Builder.Tests/BotFrameworkAdapterTests.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
+﻿using System.Security.Claims;
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.Tests/BotFrameworkAdapterTests.cs]: Copyright statement.

Comment (gabog) [reply tests/Microsoft.Bot.Builder.Tests/BotFrameworkAdapterTests.cs]: Added

**Comment by cleemullins** [comment]
```
No issues found in Microsoft.Bot.Builder.dll
No issues found in Microsoft.Bot.Builder.AI.Luis.dll
No issues found in Microsoft.Bot.Builder.AI.QnA.dll
No issues found in Microsoft.Bot.Builder.ApplicationInsights.dll
No issues found in Microsoft.Bot.Builder.Azure.dll
No issues found in Microsoft.Bot.Builder.Dialogs.dll
No issues found in Microsoft.Bot.Builder.TemplateManager.dll
No issues found in Microsoft.Bot.Configuration.dll
No issues found in Microsoft.Bot.Connector.dll
No issues found in Microsoft.Bot.Schema.dll

```
